src/qibolab/instruments/emulator/backends/generic.py

```diff
@@ -0,0 +1,315 @@
+from collections import OrderedDict
+from typing import List, Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+from IPython.display import Latex, display
+from matplotlib.figure import Figure
+
+
+def dec_to_basis_string(x: int, nlevels: list = [2]) -> list:
+    """Converts an integer to a generalized bitstring in the computation basis
+    of the full Hilbert space.
+
+    Args:
+        x (int): The integer (decimal number) to be converted.
+        nlevels (list): The list of nlevels to convert the decimal number to. Defaults to [2].
+
+    Returns:
+        list: Generalized bitstring of x.
+    """
+    nqubits = len(nlevels)
+    output_list = []
+    y = x
+    subdims_ = np.multiply.accumulate(nlevels)
+    subdims = (subdims_[-1] / subdims_).astype(int)
+
+    for sub_dim in subdims:
+        coeff = np.divmod(y, sub_dim)[0]
+        output_list.append(coeff)
+        y -= coeff * sub_dim
+
+    return output_list
+
+
+def make_comp_basis(
+    qubit_list: List[Union[int, str]], qid_nlevels_map: dict[Union[int, str], int]
+) -> np.ndarray:
+    """Generates the computational basis states of the Hilbert space.
+
+    Args:
+        qubit_list (list): List of target qubit indices to generate the local Hilbert space of the qubits that respects the order given by qubit_list.
+        qid_nlevels_map (dict): Dictionary mapping the qubit IDs given in qubit_list to their respective Hilbert space dimensions.
+
+    Returns:
+        `np.ndarray`: The list of computation basis states of the local Hilbert space in a numpy array.
+    """
+    nqubits = len(qubit_list)
+
+    qid_list = [str(qubit) for qubit in qubit_list]
+    nlevels = [qid_nlevels_map[qid] for qid in qid_list]
+    comp_basis_list = []
+    comp_dim = np.prod(nlevels)
+    for ind in range(comp_dim):
+        comp_basis_list.append(dec_to_basis_string(ind, nlevels=nlevels))
+
+    return np.array(comp_basis_list)
```

Comment:
Same as below: it doesn't seem backend specific. And indeed it is even used inside `pulse_simulator.py` in the folder above. Why is it here?

Reply:
For this I agree, I will move it to `pulse_simulator.py`. Previously `make_comp_basis` was used by the backend module but I have made changes along the way and the current version only uses it in `get_samples` in `pulse_simulator.py` so it should be there now.

Reply:
Fixed in ded2a9d0ab40915da6d610fe77ce3a422ae3ba3d